Rxmongo suffix bug fix

JeanFrancoisGuena opened the pull request:
Hello there,

I have found a bug in _RxMongoReadJournaller.scala_ file that fail my tests, even on my development machine.

I fixed it by:
* getting rid of temporary collections used in `CurrentAllPersistenceIds` class
* applying a `map` method to any `Enumerator[BSONDocument]` in `flattenCollection` methods in `CurrentAllPersistenceIds` and `CurrentAllEvents` classes (not needed in `CurrentEventsByPersistenceId` as we always deal with only one journal)
* adding a test to really check `CurrentAllEvents` behavior with suffixed collections

Also, I have replaced the deprecated `enumerate()` method, applied to cursors of `BSONDocument`, with the new `enumerator()` provided by _reactivemongo-iteratees_ module (that I have added to _build.sbt_ file)

Hope this helps, happy reviewing ;-)

scullxbones commented:
Hi @JeanFrancoisGuena -

The temporary collections are in place to support extremely large numbers of processor ids, using `distinct` will cause a document overflow @ 16MB.  The approach used here would be extremely slow with very large journals.

And thanks for making me realize I left those `println`s in, ugh.

JeanFrancoisGuena commented:
Hi @scullxbones 

I changed some timeout values inside tests, and it works better. So, temporary collections are back again, but I still have to apply the `map` method  to any `Enumerator[BSONDocument]` in `flattenCollection` methods, otherwise tests fails for suffix collections (when more than one journal is involved in fact)

I also removed your `println`s ;-)

scullxbones commented on rxmongo/src/main/scala/akka/contrib/persistence/mongodb/RxMongoReadJournaller.scala:
```diff
@@ -123,7 +127,8 @@ class CurrentAllEvents(val driver: RxMongoDriver) extends IterateeActorPublisher
       .cursor[BSONDocument]()
-      .enumerate()
+      .enumerator()
+      .map(doc => doc) // this is needed for suffix collections
```
I'm still confused by why this would be needed.  Running tests locally to try to duplicate.

JeanFrancoisGuena replied:
I cannot really explain this, but running `RxMongoSuffixReadJournalSpec` test without the `map(doc => doc)` line leads to errors like this:
```
[info] - should support the current all events query *** FAILED *** (2 seconds, 888 milliseconds)
[info]   20 was not equal to 0 (ReadJournalSpec.scala:185)
...
[info] - should support the current persistence ids query *** FAILED *** (7 seconds, 157 milliseconds)
[info]   List("1") did not contain all of ("1", "2", "3", "4", "5") (ReadJournalSpec.scala:212)
```
It seems that only the first item of enumerator returned by `driver.getJournalCollections()` is handled, and then nothing happens, the stream seems to be "stopped" or "waiting for something"...
Do you obtain the same errors ?

scullxbones replied:
I do get those exact errors.  It looks like a bug in that new rxmongo `iteratee` library.  Or at a minimum a regression in behavior from `.enumerate()`, which works without the `map(identity)` despite the deprecation warning.

Let's go back to the warnings by using the deprecated `enumerate` and remove the not-quite-ready rxmongo extension library from the build.

JeanFrancoisGuena replied:
Done !

Well, this PR should be renamed: "just add a test to really check CurrentAllEvents behavior with suffixed collections" ;-)
